filters/fadein/fadein.go

```diff
@@ -0,0 +1,252 @@
+		return "", "", err
+	}
+
+	return normalizeSchemeHost(u.Scheme, u.Host)
+}
+
+func endpointKey(scheme, host string) string {
+	return fmt.Sprintf("%s://%s", scheme, host)
+}
+
+func (endpointCreated) CreateFilter(args []interface{}) (filters.Filter, error) {
+	if len(args) != 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	e, ok := args[0].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	s, h, err := normalizeEndpoint(e)
+	if err != nil {
+		return nil, err
+	}
+
+	ec := endpointCreated{which: endpointKey(s, h)}
+	switch v := args[1].(type) {
+	case int:
+		ec.when = time.Unix(int64(v), 0)
+	case float64:
+		ec.when = time.Unix(int64(v), 0)
+	case string:
+		t, err := time.Parse(time.RFC3339, v)
+		if err != nil {
+			return nil, err
+		}
+
+		ec.when = t
+	case time.Time:
+		ec.when = v
+	default:
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	// mitigate potential flakyness caused by clock skew. When the created time is in the future based on
+	// the local clock, we ignore it.
+	now := time.Now()
+	if ec.when.After(now) {
+		log.Errorf("Endpoint created time in the future: %v. Potential clock skew.", ec.when)
+		ec.when = time.Time{}
+	}
+
+	return ec, nil
+}
+
+func (endpointCreated) Request(filters.FilterContext)  {}
+func (endpointCreated) Response(filters.FilterContext) {}
+
+func NewPostProcessor() routing.PostProcessor {
+	return &postProcessor{
+		detected: make(map[string]detectedFadeIn),
+	}
+}
+
+func (p *postProcessor) Do(r []*routing.Route) []*routing.Route {
+	const configErrFmt = "Error while processing endpoint fade-in settings: %s, %s, %v."
+	now := time.Now()
+
+	for _, ri := range r {
+		if ri.Route.BackendType != eskip.LBBackend {
+			continue
+		}
+
+		ri.LBFadeInDuration = 0
+		ri.LBFadeInEase = 1
+		endpointsCreated := make(map[string]time.Time)
+		for _, f := range ri.Filters {
+			if d, ok := f.Filter.(duration); ok {
+				ri.LBFadeInDuration = time.Duration(d)
+			}
+
+			if e, ok := f.Filter.(ease); ok {
+				ri.LBFadeInEase = float64(e)
+			}
+
+			if ec, ok := f.Filter.(endpointCreated); ok {
+				endpointsCreated[ec.which] = ec.when
+			}
+		}
+
+		if ri.LBFadeInDuration <= 0 {
+			continue
+		}
+
+		for i := range ri.LBEndpoints {
+			ep := &ri.LBEndpoints[i]
+
+			s, h, err := normalizeSchemeHost(ep.Scheme, ep.Host)
+			if err != nil {
+				log.Errorf(configErrFmt, ep.Scheme, ep.Host, err)
+				continue
+			}
+
+			key := endpointKey(s, h)
+			detected := p.detected[key].when
+			if detected.IsZero() || endpointsCreated[key].After(detected) {
+				detected = now
+			}
+
+			ep.Detected = detected
+			p.detected[key] = detectedFadeIn{
+				when:       detected,
+				duration:   ri.LBFadeInDuration,
+				lastActive: now,
+			}
+		}
+	}
+
+	// cleanup old detected, considering last active
+	for key, d := range p.detected {
+		// this allows tolerating when a fade-in endpoint temporarily disappears:
+		if d.lastActive.Add(d.duration).Before(now) {
```

Comment:
I am not sure if I understand.
The loop is to clean up the state map of the fade-in endpoints as far as I understand.
For me the natural way would be: `now > when + duration`
Would this not make more sense to have this if condition:
```
if when.Add(d.duration).After(now) { delete...}
```

Reply:
the point of considering the lastActive is to smoothen out the effect of possible flakiness, e.g. when the infrastructure temporarily doesn't detect an endpoint. By using the last active time, we can avoid resetting the fade-in unless the endpoint disappears for at least the duration of the fade-in.

Reply:
So `p.detected` stores endpoints which do not need to be passed via dataclient.
And we clean up endpoints that are hitting this case here, such that we tolerate temporary failures.
Ok got it now.